### ISSUE TITLE
CloudFront Lambda@Edge

### ISSUE THREAD

**Issue by nwhite89** [serverless/serverless#9658]
Hi all,

The problem I'm trying to solve is allowing for Cloudfront Lambda@Edge.

Currently CloudFront Lambda@Edge generation does not allow for details such as HttpVersion and Alternative domains. Currently this is supported via resources but is not supported via Cloudfront events.

```
      - cloudFront:
          eventType: viewer-response
          includeBody: true
          pathPattern: /docs*
          cachePolicy:
            # Note, you can use only one of name or id
            name: myCachePolicy1 # Refers to a Cache Policy defined in provider.cloudFront.cachePolicies
            id: 658327ea-f89d-4fab-a63d-7e88639e58f6 # Refers to any external Cache Policy id
          origin:
            DomainName: serverless.com
            OriginPath: /framework
            CustomOriginConfig:
              OriginProtocolPolicy: match-viewer
```

Is the current example options but within this to have CNAMEs and HttpVersion for example.

**Comment by pgrzesik** [comment]
Hello @nwhite89 - thanks for your proposal. How do you imagine it should be supported on this event? What properties would make the most sense? Would you be able to create a proposal that describes these properties along with possible values? It would be much appreciated :bow:

**Comment by nwhite89** [comment]
@pgrzesik Figuring out exactly how to go about this due to the nature of the behaviours is done is slightly tricky there are two slightly different options.

The first approach would be to use the `cloudfront` provider option similar to the existing `cachePolicies` https://github.com/serverless/serverless/blob/ae67bf448950b199eaf2a0ec406349d63f60424c/docs/providers/aws/events/cloudfront.md#cache-policy-configuration 

With everything that's here https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-distribution-distributionconfig.html 

The other option would be to have it on the cloudfront function handler itself where the default behaviour is set. But again with distribution config as an option.

I'd imagine this as a merge so that the current defaults that I believe can continue to be used but at the same time overriden in parts

**Comment by pgrzesik** [comment]
So, if I'm understanding correctly, it would be possible to define these properties on `provider` level and that will be inherited by all functions that use `cloudFront` events and also have the ability to override it on specific function level - correct?

**Comment by nwhite89** [comment]
@pgrzesik It would be inherited in a way. Basically any settings that is added to the CloudFront distribution is for any behaviour beneath it. Currently each function is basically a behaviour for the overall CloudFront distribution so this comes natural of inheriting from the CloudFront default similar to cachePolicies.

With regards to the properties obviously the most ideal would be all options as optional. But I would say at a minimum for this to be useful to people as a feature. 
- Aliases (basically allows people to use a custom domain for the Lambda@Edge)
- ViewerCertificate (Required for SSL with the custom domains)
- HttpVersion (I mean who doesn't want http2 these days :))
- DefaultRootObject (This for me would be a nice to have as it covers any behaviours not caught)

Now with regards to this I was thinking perhaps

```
- cloudFront:
  - DistributionConfig
    - Aliases:
      sub-1.domain.com
      sub-2.domain.com
    - HttpVersion: 'http2'
    - DefaultRootObject: '/file-name.html'
    - ViewerCertificate: 
      - SslSupportMethod: 'sni-only'
      - AcmCertificateArn: 'CERTIFICATE_ARN'
```

I'm wondering if it's done like this so it's effectively setting via the YML and in the code it takes the `.DistributionConfig` and merges with the existing way of setting DistributionConfig here https://github.com/serverless/serverless/blob/a6f4dc3b2be2aa9ba2255d57cf9a9d23eae02994/lib/plugins/aws/package/compile/events/cloudFront.js#L663

Meaning that people have the option to set everything their own without the code needing to 100% know about each individual config. But I'm not sure how you would feel about this

```
const DistributionConfig = _.merge({
            Comment: `${this.serverless.service.service} ${this.provider.getStage()}`,
            Enabled: true,
            Origins: origins,
          }, Object.assign(myDistributionConfigObj, { DefaultCacheBehavior: behaviors[0] }))
const CloudFrontDistribution = {
        Type: 'AWS::CloudFront::Distribution',
        Properties: {
          DistributionConfig: DistributionConfig,
        },
      };
```

Not the most elegant way but just to give you an idea

**Comment by pgrzesik** [comment]
Thanks a lot for explanation @nwhite89 and sorry for not responding - I was out recently and some other priorities came in the way last week so I didn't have a chance to take a deeper look at this proposal. As for accepting generic properties and just pass them as-is to CloudFormation, I'm a little bit skeptical, as it's a bit hard to provider proper schema checks for it (well, we have to specify all of them anyway in schema) and it the properties themselves are a bit different (PascalCase for CF vs camelCase which is a standard in Framework), which looks "weird" to me. It is worth noting though, that we already support such approach in one case - for configuring extra properties for S3 buckets https://www.serverless.com/framework/docs/providers/aws/events/s3/#custom-bucket-configuration.

@medikoo What is your stance on accepting such CF settings and just passing them as-is? Should we support such patterns of the `s3` one is an exception and we should rather not do that?

**Comment by medikoo** [comment]
@pgrzesik probably best solution is to open the door to extend generated CF resources, but in agnostic way (so we wouldn't validate those extensions via schema).

And we could make this event as first to support that.

(It's similar to what @mnapoli was proposing with constructs).

e.g. we could introduce a `cloudFormationExtensions` property on event, which will be a map of named resources where each item will host free form CF extensions, so something as:

```yaml
functions:
  someFunct:
    events:
      - cloudFront:
          ....
          cloudFormationExtensions:
            distribution:
              # free form not validated by schema
              Enabled: true
              Origins: ...
```

Ofc this example is not right, and should not be taken literally. Better context for such extensions in context of this event might be `provider.cloudFront`.

I would also assume that trhough those extensions only properties not generated by the Framework can be filled, and if there's a collision we crash (to avoid accidental issues)

--- 

Question is, how it'll look in context of this use case?

**Comment by mnapoli** [comment]
Hi! Yep, we are on the topic of configuring non-Lambda resources via serverless.yml.

Given we are already exploring general approaches to address this problem (e.g. constructs), it would make more sense to wait on that specific topic rather than implement a solution specifically for CloudFront.

As a workaround, is it possible to extend the CF resource like this?

```yaml
functions:
  someFunct:
    events:
      - cloudFront:
           ...

resources:
  extensions:
    <CF-logical-ID>:
      Properties:
        Enabled: true
        ...
```

**Comment by nwhite89** [comment]
@pgrzesik no problem whatsoever and makes sense. @medikoo I like the sounds of this approach the only thing that I guess it would need to be for the one only with `isDefaultOrigin: true`.

If using the CF resource approach mentioned by @mnapoli `<CF-logical-ID>` I'm assuming that's as a placeholder for now to be decided upon loater?

**Comment by mnapoli** [comment]
> I'm assuming that's as a placeholder for now to be decided upon later?

That ID should be in your generated CloudFormation template.

**Comment by medikoo** [comment]
> @pgrzesik no problem whatsoever and makes sense. @medikoo I like the sounds of this approach the only thing that I guess it would need to be for the one only with isDefaultOrigin: true.

At begining I would focus on your use case. So define which resources you want to extend, and then we need to define how to name them and in which context (whether event, or provider scope)

> If using the CF resource approach mentioned by @mnapoli <CF-logical-ID> I'm assuming that's as a placeholder for now to be decided upon later?

This works if you can statically assume a resource id that will be generated by the Framework, which I take should be possible in your case (?)

Technically this is the default general CloudFormation extension mechanism as provided by the Framework, but it's not very friendly to use, has its own specific issues

**Comment by nwhite89** [comment]
I see @medikoo and @mnapoli thanks for confirming. Regarding the ID the way I was planning / hoping to use this was having it "automatic" without requiring to know the ID per resource, which for this to work I'm right in saying you'd have to run the deploy first to find the ID? Unless I'm missing the approach properly. This was the reason I was thinking more along the lines of what is done with cache behaviours for this already.

**Comment by medikoo** [comment]
> which for this to work I'm right in saying you'd have to run the deploy first to find the ID?

You may just run `sls package` (which in most cases is just offline operation) and inspect the generated template

**Comment by nwhite89** [comment]
Ah

> > which for this to work I'm right in saying you'd have to run the deploy first to find the ID?
> 
> You may just run `sls package` (which in most cases is just offline operation) and inspect the generated template

Ah I see ok makes sense, if that's the case I think the ideal would be not needing this extra step and going via cache policy route if this is acceptable as it makes it a lot more streamlined for users.

**Comment by medikoo** [comment]
@nwhite89 We're open to accept a contribution that introduces support for something as `cloudFormationExtensions` (as proposed in comment above).

Still let's first specify exactly which resources we will allow to extend at which context (provider, event, function), and how we will name them.

**Comment by napter** [comment]
Is it currently possible to add distributionconfig properties like alias and viewercertificate in a lambda@edge cloudfront event? Or that pending on the resolution of this issue?
